resteasy-core/src/main/java/org/jboss/resteasy/plugins/providers/MultiValuedArrayParamConverter.java

```diff
@@ -0,0 +1,189 @@
+package org.jboss.resteasy.plugins.providers;
+
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Array;
+import java.lang.reflect.Method;
+
+import javax.annotation.Priority;
+import javax.ws.rs.ext.ParamConverter;
+import javax.ws.rs.ext.Provider;
+
+import org.jboss.resteasy.core.StringParameterInjector;
+import org.jboss.resteasy.resteasy_jaxrs.i18n.Messages;
+
+/**
+ * @author Marek Kopecky mkopecky@redhat.com
+ * @author Ron Sigal rsigal@redhat.com
+ */
+@Provider
+@Priority(Integer.MAX_VALUE)
+public class MultiValuedArrayParamConverter implements ParamConverter<Object>
+{
+   private Class<?> rawType;
+   private StringParameterInjector stringParameterInjector;
+   private String separator;
+
+   public MultiValuedArrayParamConverter(Class<?> rawType, Annotation[] annotations, StringParameterInjector stringParameterInjector, String separator)
+   {
+      this.rawType = rawType;
+      this.stringParameterInjector = stringParameterInjector;
+      this.separator = separator;
+   }
+
+   @Override
+   public String toString(Object value)
+   {
+      if (value == null)
+      {
+         return null;
+      }
+      if (!value.getClass().isArray())
+      {
+         throw new RuntimeException(Messages.MESSAGES.expectedArray(value.getClass().getName()));
+      }
+      Object[] values = new Object[Array.getLength(value)];
+      if (values.length == 0)
+      {
+         return "";
+      }
+      if (rawType.getComponentType().isPrimitive())
+      {
+         copyPrimitiveArray(values, value);
+      }
+      else
+      {
+         values = (Object[]) value;
+      }
+      return stringify(values);
+   }
+
+   @Override
+   public Object fromString(String param)
+   {
+      String[] elements = param.split(separator);
+      Object array = Array.newInstance(rawType.getComponentType(), elements.length);
+      try
+      {
+         for (int i = 0; i < elements.length; i++)
+         {
+            Array.set(array, i, stringParameterInjector.extractValue(elements[i]));
+         }
+      }
+      catch (Exception e)
+      {
+         throw new RuntimeException("Ooops");
+      }
+      return array;
+   }
+
+   ////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+   @SuppressWarnings({ "rawtypes", "unchecked" })
+   private String stringify(Object[] values)
+   {
+      ParamConverter converter = stringParameterInjector.getParamConverter();
+      Method saveAsString = null;
+      if (converter == null)
+      {
+         try
+         {
+            saveAsString = values[0].getClass().getMethod("saveAsString");
```

Comment:
Better look for a `stringParameterInjector.getHeaderDelegate()` before looking for `saveAsString()` method. This way, in case the parameter is a `@HeaderParam` and a `HeaderrDelegate<T>` has been registered for it let's use it.

Reply:
Ok, I'm working on it.

Reply:
Done.